### PR TITLE
Enforce unconditional attributes within method bodies

### PR DESCRIPTION
**Pull request by jcouv**
Addresses part of https://github.com/dotnet/roslyn/issues/36039 
`MaybeNullWhen` and `NotNullWhen` are treated unconditionally to produce fewer warnings.
`DoesNotReturn` and `DoesNotReturnIf` are not yet enforced.
OHI will come later.

Fixes https://github.com/dotnet/roslyn/issues/40139 (`[DisallowNull] T` parameter has not-null initial value)
Fixes https://github.com/dotnet/roslyn/issues/39922

Relates to discussion in LDM on Jan 6th 2020.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 1]
```diff
@@ -6782,11 +6827,6 @@ private Symbol VisitMemberAccess(BoundExpression node, BoundExpression receiverO
                 }
             }
 
-            if (_expressionIsRead)
-            {
-                ReportMaybeNullFromTypeParameterValueIfNeeded(node, resultType, memberAnnotations);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: :memo: removed empty method

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler @cston for review. Thanks

I'll add a note to breaking changes.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -3459,7 +3465,7 @@ private FlowAnalysisAnnotations GetParameterAnnotations(ParameterSymbol paramete
         /// Fix a TypeWithAnnotations based on Allow/DisallowNull annotations prior to a conversion or assignment.
         /// Note this does not work for nullable value types, so an additional check with <see cref="CheckDisallowedNullAssignment"/> may be required.
         /// </summary>
-        private static TypeWithAnnotations ApplyLValueAnnotations(TypeWithAnnotations declaredType, FlowAnalysisAnnotations flowAnalysisAnnotations)
+        internal static TypeWithAnnotations ApplyLValueAnnotations(TypeWithAnnotations declaredType, FlowAnalysisAnnotations flowAnalysisAnnotations)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >internal [](start = 8, length = 8)

Is this change needed?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 3]
```diff
@@ -3883,12 +3894,34 @@ private VisitArgumentResult VisitArgumentEvaluate(BoundExpression argument, RefK
...
+                // [NotNull]string -> string
+                // [NotNull]TClass -> TClass
+                // [NotNull]TClass? -> TClass
+                //
+                // While others don't:
+                // [NotNull]int? -> X
+                // [NotNull]TNullable -> X
+                // [NotNull]TStruct? -> X
+                // [NotNull]TOpen -> X
+                return type is null || (type.Kind == SymbolKind.TypeParameter && !type.IsReferenceType) || type.IsNullableTypeOrTypeParameter();
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >type is null [](start = 23, length = 12)

When do we hit this, and is it necessary to report a warning in that case?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 4]
```diff
@@ -6099,6 +6129,21 @@ static FlowAnalysisAnnotations getPropertyAnnotations(SourcePropertySymbol prope
             }
         }
 
+        FlowAnalysisAnnotations ToInwardAnnotations(FlowAnalysisAnnotations outwardAnnotations)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: `private static`

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 5]
```diff
@@ -5439,10 +5439,10 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
     <value>Argument cannot be used as an output for parameter due to differences in the nullability of reference types.</value>
   </data>
   <data name="WRN_DisallowNullAttributeForbidsMaybeNullAssignment" xml:space="preserve">
-    <value>A possible null value may not be assigned to a target marked with the [DisallowNull] attribute</value>
+    <value>A possible null value may not be assigned to a target marked with the [DisallowNull] attribute or output to a target marked with the [NotNull] attribute</value>
```
Comment (cston) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: >output to a target marked with [](start = 109, length = 30)

Is this qualifier necessary? (Is it necessary to differentiate "assigned to" and "output to"? Also, "output to" feels like "yield".)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 6]
```diff
@@ -27636,6 +27710,358 @@ void Main(string? s)
...
+        public void UnconditionalAttributesAffectPropertyBodies_ReferenceType_AutoProp()
+        {
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    [DisallowNull] string? P1 { get; set; } = """";
+    [AllowNull] string P2 { get; set; } = """";
+    [MaybeNull] string P3 { get; set; } = """";
+    [NotNull] string? P4 { get; set; } = """";
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: It's not clear we're that the attributes affect the property bodies since there are no warnings. Should the assignments be changed to ` = null'` to at least test two of the four?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 7]
```diff
@@ -30937,11 +31435,11 @@ public void MaybeNull_ReturnValue_02_WithNotNull()
 @"using System.Diagnostics.CodeAnalysis;
 class Program
 {
-    [return: MaybeNull, NotNull] static T F1<T>(T t) => t;
+    [return: MaybeNull, NotNull] static T F1<T>(T t) => t; //
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >// [](start = 59, length = 2)

Are these empty comments needed?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 8]
```diff
@@ -119893,12 +120459,44 @@ class Program
...
-                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "x").WithLocation(9, 71));
+                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "x").WithLocation(9, 71),
+                // (14,71): warning CS8607: A possible null value may not be assigned to a target marked with the [DisallowNull] attribute or output to a target marked with the [NotNull] attribute
+                //     static void FA<T>(              T x,      [NotNull]out T y) { y = x; } // 5
+                Diagnostic(ErrorCode.WRN_DisallowNullAttributeForbidsMaybeNullAssignment, "x").WithLocation(14, 71));
+        }
+
+        [Fact]
+        [WorkItem(36039, "https://github.com/dotnet/roslyn/issues/36039")]
+        public void NontNullOutParameterWithVariousTypes()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >Nont [](start = 20, length = 4)

Typo

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 9]
```diff
@@ -32549,7 +33072,14 @@ static void M1<T>(T t1)
...
 
             var comp2 = CreateNullableCompilation(new[] { source, lib_cs, NotNullAttributeDefinition });
-            comp2.VerifyDiagnostics();
+            comp2.VerifyDiagnostics(
+                // (4,32): warning CS8607: A possible null value may not be assigned to a target marked with the [DisallowNull] attribute or output to a target marked with the [NotNull] attribute
+                //     [NotNull]public TOpen P1 { get; set; } = default!;
+                Diagnostic(ErrorCode.WRN_DisallowNullAttributeForbidsMaybeNullAssignment, "get;").WithLocation(4, 32),
+                // (14,35): warning CS8607: A possible null value may not be assigned to a target marked with the [DisallowNull] attribute or output to a target marked with the [NotNull] attribute
+                //     [NotNull]public TStruct? P5 { get; set; }
+                Diagnostic(ErrorCode.WRN_DisallowNullAttributeForbidsMaybeNullAssignment, "get;").WithLocation(14, 35)
```
Comment (sharwell) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: 💡 It would be good to have a test showing that the value can be assigned in a constructor to eliminate the warning (may already exist as a test).

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 10]
```diff
@@ -33056,15 +33594,15 @@ public void NotNull_RefParameter_03()
 class Program
 {
     static void F1([NotNull]ref string t) { t = null; } // 0
-    static void F2([NotNull]ref string? t) { t = null; }
+    static void F2([NotNull]ref string? t) { t = null; } // 1
```
Comment (sharwell) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: 💡 Consider adding `F3` with `t = null!;` to show that suppressions work here.

```suggestion
    static void F2([NotNull]ref string? t) { t = null; } // 1
    static void F3([NotNull]ref string? t) { t = null!; }
```

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 11]
```diff
@@ -27636,6 +27710,358 @@ void Main(string? s)
...
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+public class C<T>
+{
+    void M1([AllowNull]T x)
+    {
+        x.ToString(); // 1
+    }
+
+    void M2([DisallowNull]T x)
```
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: @sharwell I disagree with this one. `x = default` is disallowed in absence of `[DisallowNull]`, so should still be disallowed.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 12]
```diff
@@ -28095,7 +28521,14 @@ static void M1<T>([MaybeNull]T t1)
     }
 }";
             var comp = CreateNullableCompilation(new[] { source, AllowNullAttributeDefinition, NotNullAttributeDefinition, MaybeNullAttributeDefinition });
-            comp.VerifyDiagnostics();
+            comp.VerifyDiagnostics(
+                // (4,43): warning CS8607: A possible null value may not be assigned to a target marked with the [DisallowNull] attribute or output to a target marked with the [NotNull] attribute
+                //     [AllowNull, NotNull]public TOpen P1 { get; set; } = default!;
+                Diagnostic(ErrorCode.WRN_DisallowNullAttributeForbidsMaybeNullAssignment, "get;").WithLocation(4, 43),
```
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: We analyze a body that says `return backingField;`. But that backing field is a `TOpen` and lacks attribute annotations. So we warn.
I'll take a look to see if we can refine that.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 13]
```diff
@@ -28095,7 +28521,14 @@ static void M1<T>([MaybeNull]T t1)
     }
 }";
             var comp = CreateNullableCompilation(new[] { source, AllowNullAttributeDefinition, NotNullAttributeDefinition, MaybeNullAttributeDefinition });
-            comp.VerifyDiagnostics();
+            comp.VerifyDiagnostics(
```
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: This is existing behavior. The return value for `P3` is `[NotNull]TStruct?` so no nullable warning.
I think we don't differentiate on auto-props, although presumably we could.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 14]
```diff
@@ -29930,8 +30422,8 @@ public class C
 }
 ";
 
-            // Honor nullability attributes in optional parameters  https://github.com/dotnet/roslyn/issues/36039
             var comp = CreateNullableCompilation(new[] { AllowNullAttributeDefinition, source });
+            // Note: fixing this was tricky as it involved a cycle (FlowAnnotations, attribute decoding, default value)
```
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Filed https://github.com/dotnet/roslyn/issues/40975

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: I find out that some warnings involving backing fields and other compiler-generated values were suppressed. I removed that, which not only keeps this warning on `P1`, but also adds a warning on `P3`.
@sharwell @cston Let me know what you think about those. In my opinion, those are bad usages of the attributes, and so I think it's actually desirable to warn.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 15]
```diff
@@ -3909,6 +3908,11 @@ private void CheckDisallowedNullAssignment(TypeWithState state, FlowAnalysisAnno
 
             static bool hasNoNonNullableCounterpart(TypeSymbol type)
             {
+                if (type is null)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: :memo: not sure if we can hit this. Just included for safety.

**Comment by jcouv** [comment]
@cston @sharwell Addressed your feedback

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 16]
```diff
@@ -27636,6 +27710,358 @@ void Main(string? s)
...
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    void M1([AllowNull]string x)
+    {
+        x.ToString(); // 1
+    }
+
+    void M2([DisallowNull]string? x)
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: ➡️ Test was added

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 17]
```diff
@@ -27636,6 +27710,358 @@ void Main(string? s)
...
+            var source = @"
+using System.Diagnostics.CodeAnalysis;
+public class C
+{
+    void M1([AllowNull]int? x)
+    {
+        x.Value.ToString(); // 1
+    }
+
+    void M2([DisallowNull]int? x)
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: ➡️ Test was added

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: That makes sense to me. Consider adding a line showing this diagnostic _is_ reported for the case.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Both cases can be resolved by removing the unnecessary initializing expression. If users find this confusing, we can iterate on it.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: It looks like there now _is_ a warning for P3, so it appears more consistent at least. 😄

[the remark this answers is not in the document]
Comment (sharwell) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: ➡️ A test has now been added.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 18]
```diff
@@ -28824,6 +29257,53 @@ public class Derived : Base
...
+        _f = t;
+    }
+}
+";
+            var comp = CreateNullableCompilation(new[] { DisallowNullAttributeDefinition, source });
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact, WorkItem(39922, "https://github.com/dotnet/roslyn/issues/39922")]
+        public void EnforcedInMethodBody_Misc()
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: ➡️ Test was added

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: ➡️ Link restored/updated

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 9 above)
Comment (sharwell) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: ❔ I didn't see a test that used the following in a constructor:

```
P5 = new TStruct();
```

Unlike the newly-added `NotNull_Property_InitializedInConstructor` cases, I would expect the above to not produce a warning _and_ resolve the warning reported on the property definition in the two tests here.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 10 above)
Comment (sharwell) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: ➡️ Test was added

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 19]
```diff
@@ -27943,13 +27954,31 @@ public class C
     [DisallowNull] string? P1 { get; set; } = """";
     [AllowNull] string P2 { get; set; } = """";
     [MaybeNull] string P3 { get; set; } = """";
-    [NotNull] string? P4 { get; set; } = """";
+    [NotNull] string? P4 { get; set; } = """"; // 1
+
+    [DisallowNull] string? P5 { get; set; } = null; // 2
+    [AllowNull] string P6 { get; set; } = null;
```
Comment (sharwell) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: 😕 I would expect a warning for `P6`, because this is essentially the same as `P3` in test `AllowNull_Property_WithNotNull` (where a warning is reported)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: From discussion with @cston I'm going to restore the logic to skip analyzing bodies of auto-prop getters and setters. We'll lose some warning about misused attributes, but that's okay.

One reason is that there is a scenario where that would produce an unwanted warning: `[MaybeNull, AllowNull] T { get; set; }` which the user can't fix.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: I think that case is already covered. See `DefaultLiteralInConditional`

**Comment by jcouv** [comment]
@cston @sharwell Please take another look. Thanks

Note: I reverted the analysis of synthesized code. We'll have to report misused attributes some other way or at least in a different PR. The difficult scenario is `[MaybeNull, AllowNull] T { get; set; }`.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 20]
```diff
@@ -3883,12 +3894,40 @@ private VisitArgumentResult VisitArgumentEvaluate(BoundExpression argument, RefK
             Debug.Assert(!this.IsConditionalState);
         }
 
-        private void CheckDisallowedNullAssignment(TypeWithState state, FlowAnalysisAnnotations annotations, Location location)
+        private void CheckDisallowedNullAssignment(TypeWithState state, FlowAnalysisAnnotations annotations, Location location, BoundExpression boundValueOpt = null)
         {
-            if (((annotations & FlowAnalysisAnnotations.DisallowNull) != 0) && state.Type.IsNullableTypeOrTypeParameter() && state.MayBeNull)
+            if (boundValueOpt is object && boundValueOpt.WasCompilerGenerated)
+            {
+                // We need to skip `return backingField;` in auto-prop getters
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >// We need to skip `return backingField;` in auto-prop getters [](start = 16, length = 62)

Can we simply avoid running `NullableWalker` on auto-property accessors?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 21]
```diff
@@ -3883,12 +3894,40 @@ private VisitArgumentResult VisitArgumentEvaluate(BoundExpression argument, RefK
             Debug.Assert(!this.IsConditionalState);
         }
 
-        private void CheckDisallowedNullAssignment(TypeWithState state, FlowAnalysisAnnotations annotations, Location location)
+        private void CheckDisallowedNullAssignment(TypeWithState state, FlowAnalysisAnnotations annotations, Location location, BoundExpression boundValueOpt = null)
         {
-            if (((annotations & FlowAnalysisAnnotations.DisallowNull) != 0) && state.Type.IsNullableTypeOrTypeParameter() && state.MayBeNull)
+            if (boundValueOpt is object && boundValueOpt.WasCompilerGenerated)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >boundValueOpt is object && boundValueOpt.WasCompilerGenerated [](start = 16, length = 61)

Consider using `?.`

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 20 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: We could, but I'd prefer to contain the scope of this PR given timeline.
Also, even if we skip auto-prop accessors, we'll still need some logic to skip compiler-generated nodes because of `field = initializerValue;` that is injected to constructor body and analyzed.

---
In reply to: [367228167](https://github.com/dotnet/roslyn/pull/40789#discussion_r367228167) [](ancestors = 367228167)

Comment (cston) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Please consider logging a bug to skip auto-property accessors.

---
In reply to: [367632535](https://github.com/dotnet/roslyn/pull/40789#discussion_r367632535) [](ancestors = 367632535,367228167)

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for a second review please.
Thanks @cston!

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 20 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Filed https://github.com/dotnet/roslyn/issues/41017

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 22]
```diff
@@ -3883,12 +3894,40 @@ private VisitArgumentResult VisitArgumentEvaluate(BoundExpression argument, RefK
             Debug.Assert(!this.IsConditionalState);
         }
 
-        private void CheckDisallowedNullAssignment(TypeWithState state, FlowAnalysisAnnotations annotations, Location location)
+        private void CheckDisallowedNullAssignment(TypeWithState state, FlowAnalysisAnnotations annotations, Location location, BoundExpression boundValueOpt = null)
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Consider naming `CheckDisallowNullAssignment` to more strongly denote that this is about DisallowNullAttribute

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 23]
```diff
@@ -6099,6 +6135,21 @@ static FlowAnalysisAnnotations getPropertyAnnotations(SourcePropertySymbol prope
             }
         }
 
+        private static FlowAnalysisAnnotations ToInwardAnnotations(FlowAnalysisAnnotations outwardAnnotations)
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >ToInwardAnnotations [](start = 47, length = 19)

Consider naming ToInboundAnnotations to resemble similar terminology used in VisitArguments

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 24]
```diff
@@ -6099,6 +6135,21 @@ static FlowAnalysisAnnotations getPropertyAnnotations(SourcePropertySymbol prope
...
+        {
+            var annotations = FlowAnalysisAnnotations.None;
+            if ((outwardAnnotations & FlowAnalysisAnnotations.MaybeNull) != 0)
+            {
+                // MaybeNull and MaybeNullWhen count as MaybeNull
+                annotations |= FlowAnalysisAnnotations.AllowNull;
+            }
+            if ((outwardAnnotations & FlowAnalysisAnnotations.NotNull) == FlowAnalysisAnnotations.NotNull)
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Style between this and the above `if` feels inconsistent

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 25]
```diff
@@ -119937,12 +120682,12 @@ class Program
                 // (7,71): warning CS8601: Possible null reference assignment.
                 //     static void F3<T>(   [AllowNull]T x, [DisallowNull]ref T y) { y = x; } // 3
                 Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "x").WithLocation(7, 71),
-                // (8,71): warning CS8601: Possible null reference assignment.
-                //     static void F4<T>(   [AllowNull]T x,    [MaybeNull]ref T y) { y = x; }
-                Diagnostic(ErrorCode.WRN_NullReferenceAssignment, "x").WithLocation(8, 71),
                 // (9,71): warning CS8601: Possible null reference assignment.
                 //     static void F5<T>(   [AllowNull]T x,      [NotNull]ref T y) { y = x; } // 4
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >T x [](start = 55, length = 3)

I think this is maybe innate to the problem, but it feels odd to have a different message here depending on whether `AllowNull` was used. In F5, T is "definitely maybe null", but in FA, T is "possibly maybe null".

**Comment by RikkiGibson** [comment]
>     static void F2<T>(   [AllowNull]T x,    [AllowNull]out T y) { y = x; } // 2

It feels like `AllowNull/DisallowNull` on `out` parameters is something that could never be what the user intended. Related to #36073

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:119870 in 501855b. [](commit_id = 501855bdc7f4d1c775a370b312dcf80285d62929, deletion_comment = True)

**Comment by RikkiGibson** [comment]
>     static void F8<T>(              T x, [DisallowNull]T y) { y = x; }

It feels like if assigning maybe-null to a `[NotNull]T` parameter gives a warning, assigning maybe-null to a `[DisallowNull]T` parameter inside the method should give a warning.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:119847 in 501855b. [](commit_id = 501855bdc7f4d1c775a370b312dcf80285d62929, deletion_comment = True)

**Comment by RikkiGibson** [comment]
>     static void FA<T>(              T x,      [NotNull]T y) { y = x; }

Also consider a `// 1` comment here.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:120557 in 501855b. [](commit_id = 501855bdc7f4d1c775a370b312dcf80285d62929, deletion_comment = False)

**Comment by RikkiGibson** [comment]
>     [return: NotNull] public override T F1<T>(T t) => t;

Consider adding a `// 1` comment here

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:33885 in 501855b. [](commit_id = 501855bdc7f4d1c775a370b312dcf80285d62929, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 26]
```diff
@@ -32552,6 +33136,110 @@ static void M1<T>(T t1)
...
+        }
+
+        [Fact, WorkItem(36073, "https://github.com/dotnet/roslyn/issues/36073")]
+        public void NotNull_Property_InitializedInConstructor_WithValues()
+        {
+            var lib_cs =
+@"using System.Diagnostics.CodeAnalysis;
+public class COpen<TOpen> where TOpen : new()
+{
+    [NotNull]public TOpen P1 { get; set; }
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: Is this just additional coverage, or did your change affect behavior of these scenarios? It doesn't seem like enforcing unconditional attributes inside the members using the attributes affects this?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 27]
```diff
@@ -30937,11 +31496,11 @@ public void MaybeNull_ReturnValue_02_WithNotNull()
 @"using System.Diagnostics.CodeAnalysis;
 class Program
 {
-    [return: MaybeNull, NotNull] static T F1<T>(T t) => t;
+    [return: MaybeNull, NotNull] static T F1<T>(T t) => t; //
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: >// [](start = 59, length = 2)

Was typing `// -1` a bridge too far? :)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 28]
```diff
@@ -28877,6 +29417,9 @@ static void M1<T>(T t1)
 }";
             var comp = CreateCompilation(new[] { DisallowNullAttributeDefinition, source }, options: WithNonNullTypesTrue());
             comp.VerifyDiagnostics(
+                // (11,12): warning CS8607: A possible null value may not be assigned to a target marked with the [DisallowNull] attribute
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: nit: Consider updating the diagnosic message in this comment. I had started to comment about how this message could be better but it turned out you already changed it for the better :)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 24 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: Yes, that's on purpose. MaybeNull is two bits, either of which should count (see comment above).
NotNull is also two bits, but we are lenient (we ignore NotNullWhenTrue and NotNullWhenFalse on their own here).

---
In reply to: [368161275](https://github.com/dotnet/roslyn/pull/40789#discussion_r368161275) [](ancestors = 368161275)

Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: I'll add a comment

---
In reply to: [368183067](https://github.com/dotnet/roslyn/pull/40789#discussion_r368183067) [](ancestors = 368183067,368161275)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 25 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: We could have the new check (from `CheckDisallowedNullAssignment`) report the same diagnostic as the regular checks that `VisitConversion` can report. But since we have more context/information here, I figured out that some hint would be helpful, when possible. But it is a bit inconsistent, I agree.

---
In reply to: [368176797](https://github.com/dotnet/roslyn/pull/40789#discussion_r368176797) [](ancestors = 368176797)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 26 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: For a while, this PR include analyzing the body of the auto-prop accessors. So those tests were useful.
Since I added the tests, I'd rather keep them even though they are less useful now :-)

---
In reply to: [368177028](https://github.com/dotnet/roslyn/pull/40789#discussion_r368177028) [](ancestors = 368177028)

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs (hunk 27 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs]: I could not bring myself to do it :-P

---
In reply to: [368177057](https://github.com/dotnet/roslyn/pull/40789#discussion_r368177057) [](ancestors = 368177057)

**Comment by jcouv** [comment]
>     static void F2<T>(   [AllowNull]T x,    [AllowNull]out T y) { y = x; } // 2

Per our discussion and spec, `NotNull` is definitely useful on inputs.
I can't think of a scenario where `AllowNull` or `DisallowNull` would be useful on an output. But I don't want to produce individual warnings for those situations until we've decided that we want to warn on misused attributes as a whole (there are many more cases that could warrant a warning).

---
In reply to: [575833532](https://github.com/dotnet/roslyn/pull/40789#issuecomment-575833532) [](ancestors = 575833532)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:119870 in 501855b. [](commit_id = 501855bdc7f4d1c775a370b312dcf80285d62929, deletion_comment = True)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 22 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: I think that "check disallowed null assignment" rings better as an action. I'll keep as-is

---
In reply to: [368161189](https://github.com/dotnet/roslyn/pull/40789#discussion_r368161189) [](ancestors = 368161189)

**Comment by jcouv** [comment]
FYI @cston, I took Rikki's suggestion so that `void M([DisallowNull]T x) { x = null; }` now produces a warning and `void M([AllowNull]string x) { x = null; }` no longer does.

**Comment by jcouv** [comment]
> FYI @cston, I took Rikki's suggestion so that void M([DisallowNull]T x) { x = null; } now produces a warning and void M([AllowNull]string x) { x = null; } no longer does.

Never mind, that won't work because of `void M([DisallowNull] ref T x) { x = null; }` scenario. I'll revert.
